File: src/screens/ServiceScreen.js
import React, { useMemo, useState } from "react";
import { View, Text, ScrollView, TextInput } from "react-native";
import { Card, Label, Field, OptionList, Action, CheckRow, useC } from "../components/ui";
import { TYPE, SPACE } from "../theme";
import { uid, getServicePackages, emptyFieldValues } from "../data/templates";
import TopBar from "../components/TopBar";
import { PhotoStrip } from "../components/Photos";

function todayISO() {
  return new Date().toISOString().slice(0, 10);
}

export default function ServiceScreen({ vehicle, packageId, onBack, onSave }) {
  const c = useC();
  const pack = useMemo(
    () => getServicePackages(vehicle).find((p) => p.id === packageId),
    [vehicle, packageId]
  );

  const [date, setDate] = useState(todayISO());
  const [odo, setOdo] = useState(String(vehicle.odometer));
  const [note, setNote] = useState("");
  const [done, setDone] = useState({});
  const [values, setValues] = useState({});
  const [photos, setPhotos] = useState([]);

  if (!pack) {
    return (
      <View style={{ flex: 1, backgroundColor: c.bg }}>
        <TopBar title="Service" onBack={onBack} />
      </View>
    );
  }

  const toggle = (cat) => {
    const turningOn = !done[cat.id];
    setDone((d) => ({ ...d, [cat.id]: turningOn }));
    setValues((v) => {
      if (!turningOn) {
        const next = { ...v };
        delete next[cat.id];
        return next;
      }
      return { ...v, [cat.id]: emptyFieldValues(cat, true) };
    });
  };

  const setField = (catId, label, val) => {
    setValues((v) => ({ ...v, [catId]: { ...(v[catId] || {}), [label]: val } }));
  };

  const checked = pack.categories.filter((cat) => done[cat.id]);

  return (
    <View style={{ flex: 1, backgroundColor: c.bg }}>
      <TopBar title={pack.name} onBack={onBack} />
      <ScrollView
        contentContainerStyle={{ paddingHorizontal: SPACE.side, paddingTop: 12, paddingBottom: 40 }}
        keyboardShouldPersistTaps="handled"
      >
        <Card>
          <Field label="Date" value={date} onChangeText={setDate} placeholder="YYYY-MM-DD" />
          <Field label="Odometer" value={odo} onChangeText={setOdo} unit="km" keyboardType="numeric" last />
        </Card>

        <View style={{ marginTop: SPACE.block }}>
          <Label style={{ marginBottom: 10 }}>Checklist</Label>
          <Card>
            {pack.categories.map((cat, i) => {
              const on = !!done[cat.id];
              const isLast = i === pack.categories.length - 1;
              const inline = cat.fields.filter((f) => f.type !== "select");
              const selects = cat.fields.filter((f) => f.type === "select");
              const showFields = on && cat.fields.length > 0;
              const catValues = values[cat.id] || {};

              return (
                <View key={cat.id}>
                  <CheckRow label={cat.name} checked={on} onToggle={() => toggle(cat)} last={showFields || isLast} />
                  {showFields &&
                    inline.map((f, fi) =>
                      f.type === "checkbox" ? (
                        <CheckRow
                          key={f.id}
                          label={f.label}
                          checked={!!catValues[f.label]}
                          onToggle={() => setField(cat.id, f.label, !catValues[f.label])}
                          last={isLast && fi === inline.length - 1 && selects.length === 0}
                        />
                      ) : (
                        <Field
                          key={f.id}
                          label={f.label}
                          unit={f.unit}
                          value={catValues[f.label] || ""}
                          onChangeText={(val) => setField(cat.id, f.label, val)}
                          keyboardType={f.type === "number" ? "numeric" : "default"}
                          placeholder={f.unit ? "e.g. 0.6" : ""}
                          last={isLast && fi === inline.length - 1 && selects.length === 0}
                        />
                      )
                    )}
                  {showFields &&
                    selects.map((f, si) => (
                      <View key={f.id}>
                        <Text style={[TYPE.meta, { color: c.muted, paddingHorizontal: SPACE.cardPad, paddingTop: 10 }]}>
                          {f.label}
                        </Text>
                        <OptionList
                          options={f.options}
                          value={catValues[f.label] || ""}
                          onChange={(val) => setField(cat.id, f.label, val)}
                        />
                        {!(isLast && si === selects.length - 1) ? (
                          <View
                            style={{
                              height: 1,
                              backgroundColor: c.hairline,
                              marginHorizontal: SPACE.cardPad,
                            }}
                          />
                        ) : null}
                      </View>
                    ))}
                </View>
              );
            })}
          </Card>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <Label style={{ marginBottom: 10 }}>Note</Label>
          <Card>
            <TextInput
              value={note}
              onChangeText={setNote}
              placeholder="Optional — applies to every item"
              placeholderTextColor={c.faint}
              multiline
              style={[
                TYPE.body,
                {
                  color: c.ink,
                  paddingHorizontal: SPACE.cardPad,
                  paddingVertical: SPACE.fieldY,
                  minHeight: 64,
                  textAlignVertical: "top",
                },
              ]}
            />
          </Card>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <Label style={{ marginBottom: 10 }}>Photos</Label>
          <Card style={{ padding: SPACE.cardPad }}>
            <PhotoStrip uris={photos} editable onChange={setPhotos} />
          </Card>
        </View>

        <Action
          label={checked.length ? `Save · ${checked.length}` : "Tick items first"}
          onPress={() => {
            if (!checked.length) return;
            const odometer = Number(odo) || 0;
            onSave(
              checked.map((cat) => ({
                id: uid(),
                categoryId: cat.id,
                categoryName: cat.name,
                date,
                odometer,
                values: values[cat.id] || {},
                note,
                service: pack.name,
                photos,
              }))
            );
          }}
        />
      </ScrollView>
    </View>
  );
}

File: lib/screens/service_screen.dart
import 'package:flutter/material.dart';

import '../models.dart';
import '../theme.dart';
import '../widgets/photos.dart';
import '../widgets/ui.dart';

class ServiceScreen extends StatefulWidget {
  final Vehicle vehicle;
  final String packageId;
  final VoidCallback onBack;
  final ValueChanged<List<LogEntry>> onSave;

  const ServiceScreen({super.key, required this.vehicle, required this.packageId, required this.onBack, required this.onSave});

  @override
  State<ServiceScreen> createState() => _ServiceScreenState();
}

class _ServiceScreenState extends State<ServiceScreen> {
  late String _date;
  late String _odo;
  String _note = '';
  final Map<String, bool> _done = {};
  final Map<String, Map<String, Object?>> _values = {};
  List<String> _photos = [];

  @override
  void initState() {
    super.initState();
    _date = todayIso();
    _odo = widget.vehicle.odometer.round().toString();
  }

  void _toggle(Category cat) {
    final turningOn = !(_done[cat.id] ?? false);
    setState(() {
      _done[cat.id] = turningOn;
      if (turningOn) {
        _values[cat.id] = emptyFieldValues(cat, true);
      } else {
        _values.remove(cat.id);
      }
    });
  }

  void _setField(String catId, String label, Object? value) {
    setState(() => _values[catId] = {...(_values[catId] ?? {}), label: value});
  }

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final pack = getServicePackages(widget.vehicle).where((p) => p.id == widget.packageId).firstOrNull;

    if (pack == null) {
      return Container(color: c.bg, child: Column(children: [TopBar(title: 'Service', onBack: widget.onBack)]));
    }

    final checked = pack.categories.where((cat) => _done[cat.id] == true).toList();

    return Container(
      color: c.bg,
      child: Column(
        children: [
          TopBar(title: pack.name, onBack: widget.onBack),
          Expanded(
            child: ListView(
              padding: const EdgeInsets.fromLTRB(AppSpace.side, 12, AppSpace.side, 40),
              keyboardDismissBehavior: ScrollViewKeyboardDismissBehavior.onDrag,
              children: [
                AppCard(children: [
                  AppField(label: 'Date', value: _date, placeholder: 'YYYY-MM-DD', onChanged: (v) => setState(() => _date = v)),
                  AppField(
                    label: 'Odometer',
                    value: _odo,
                    unit: 'km',
                    keyboardType: TextInputType.numberWithOptions(decimal: true),
                    last: true,
                    onChanged: (v) => setState(() => _odo = v),
                  ),
                ]),
                const SizedBox(height: AppSpace.block),
                const Label('Checklist', margin: EdgeInsets.only(bottom: 10)),
                AppCard(children: _buildChecklistRows(pack.categories)),
                const SizedBox(height: AppSpace.block),
                const Label('Note', margin: EdgeInsets.only(bottom: 10)),
                AppCard(
                  children: [
                    TextField(
                      onChanged: (v) => _note = v,
                      maxLines: null,
                      minLines: 3,
                      style: AppTypography.body.copyWith(color: c.ink),
                      decoration: InputDecoration(
                        isDense: true,
                        contentPadding: const EdgeInsets.symmetric(horizontal: AppSpace.cardPad, vertical: AppSpace.fieldY),
                        border: InputBorder.none,
                        hintText: 'Optional — applies to every item',
                        hintStyle: AppTypography.body.copyWith(color: c.faint),
                      ),
                    ),
                  ],
                ),
                const SizedBox(height: AppSpace.block),
                const Label('Photos', margin: EdgeInsets.only(bottom: 10)),
                AppCard(
                  padding: const EdgeInsets.all(AppSpace.cardPad),
                  children: [PhotoStrip(uris: _photos, editable: true, onChange: (p) => setState(() => _photos = p))],
                ),
              ],
            ),
          ),
          Container(
            padding: const EdgeInsets.fromLTRB(AppSpace.side, 14, AppSpace.side, 14),
            decoration: BoxDecoration(color: c.bg, border: Border(top: BorderSide(color: c.hairline, width: AppSpace.hairline))),
            child: PrimaryButton(
              label: checked.isNotEmpty ? 'Save · ${checked.length}' : 'Tick items first',
              onPressed: checked.isEmpty
                  ? null
                  : () {
                      final odometer = double.tryParse(_odo) ?? 0;
                      widget.onSave([
                        for (final cat in checked)
                          LogEntry(
                            categoryId: cat.id,
                            categoryName: cat.name,
                            date: _date,
                            odometer: odometer,
                            values: _values[cat.id] ?? {},
                            note: _note,
                            service: pack.name,
                            photos: _photos,
                          ),
                      ]);
                    },
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> _buildChecklistRows(List<Category> categories) {
    final rows = <Widget>[];
    for (var ci = 0; ci < categories.length; ci++) {
      final cat = categories[ci];
      final isLastCategory = ci == categories.length - 1;
      final on = _done[cat.id] == true;
      final showFields = on && cat.fields.isNotEmpty;
      final catValues = _values[cat.id] ?? {};

      rows.add(CheckRow(label: cat.name, checked: on, onToggle: () => _toggle(cat), last: showFields || isLastCategory));

      if (showFields) {
        final inline = cat.fields.where((f) => f.type != 'select').toList();
        final selects = cat.fields.where((f) => f.type == 'select').toList();

        for (var fi = 0; fi < inline.length; fi++) {
          final f = inline[fi];
          final isLastInline = isLastCategory && fi == inline.length - 1 && selects.isEmpty;
          if (f.type == 'checkbox') {
            rows.add(CheckRow(
              label: f.label,
              checked: catValues[f.label] == true,
              onToggle: () => _setField(cat.id, f.label, !(catValues[f.label] == true)),
              last: isLastInline,
            ));
          } else {
            rows.add(AppField(
              label: f.label,
              unit: f.unit,
              value: (catValues[f.label] as String?) ?? '',
              onChanged: (v) => _setField(cat.id, f.label, v),
              keyboardType: f.type == 'number' ? TextInputType.numberWithOptions(decimal: true) : TextInputType.text,
              placeholder: f.unit != null ? 'e.g. 0.6' : '',
              last: isLastInline,
            ));
          }
        }

        for (var si = 0; si < selects.length; si++) {
          final f = selects[si];
          final isLastSelect = isLastCategory && si == selects.length - 1;
          rows.add(Padding(
            padding: const EdgeInsets.fromLTRB(AppSpace.cardPad, 10, AppSpace.cardPad, 0),
            child: Builder(builder: (context) => AppText.meta(f.label, color: AppColors.of(context).muted)),
          ));
          rows.add(OptionList<String>(
            options: f.options ?? const [],
            value: catValues[f.label] as String?,
            onChanged: (v) => _setField(cat.id, f.label, v),
            getLabel: (v) => v,
          ));
          if (!isLastSelect) rows.add(const AppDivider());
        }
      }
    }
    return rows;
  }
}
